testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionConcurrencyTest.java

```diff
@@ -110,6 +109,10 @@ public void testConcurrentNotesChange() {
 
                     cSession.setNote(OIDCLoginProtocol.STATE_PARAM, "state-" + n);
 
+                    if (isHotRodStore) {
+                        releaseLockOnTransactionCommit(session, SYNC_USESSION);
```

Comment:
FYI: This was moved because the user session transaction is `enlistedAfterCompletion` and we need to release the lock after this transaction is committed. As `KeycloakTransactionManager` is using `List` moving this here means we release the lock after the user session transaction is committed. This is not nice but, this is only a temporary solution as it will be removed with HotRod transaction introduction.

Maybe we should consider using `enlist` for user session transactions (after we use HR transaction in HotRod store). if I remember correctly `afterCompletion` was used due to legacy Infinispan store.

Reply:
This sounds like a good plan. Could you introduce a new issue for that?

Reply:
Created https://github.com/keycloak/keycloak/issues/13575